url748e34f967_Diab1o_java_7z_archiver_tgz-pJ8-com_swemel_sevenzip_TestJ8/src/com/swemel/sevenzip/compression/lzma/Encoder.java

```diff
@@ -0,0 +1,1316 @@
+                _literalEncoder.getSubCoder(position, _previousByte).getPrice(!Base.stateIsCharState(state), matchByte, currentByte);
+        _optimum[1].makeAsChar();
+
+        int matchPrice = com.swemel.sevenzip.compression.rangecoder.Encoder.getPrice1(_isMatch[(state << Base.kNumPosStatesBitsMax) + posState]);
+        int repMatchPrice = matchPrice + com.swemel.sevenzip.compression.rangecoder.Encoder.getPrice1(_isRep[state]);
+
+        if (matchByte == currentByte) {
+            int shortRepPrice = repMatchPrice + getRepLen1Price(state, posState);
+            if (shortRepPrice < _optimum[1].Price) {
+                _optimum[1].Price = shortRepPrice;
+                _optimum[1].makeAsShortRep();
+            }
+        }
+
+        int lenEnd = ((lenMain >= repLens[repMaxIndex]) ? lenMain : repLens[repMaxIndex]);
+
+        if (lenEnd < 2) {
+            backRes = _optimum[1].BackPrev;
+            return 1;
+        }
+
+        _optimum[1].PosPrev = 0;
+
+        _optimum[0].Backs0 = reps[0];
+        _optimum[0].Backs1 = reps[1];
+        _optimum[0].Backs2 = reps[2];
+        _optimum[0].Backs3 = reps[3];
+
+        int len = lenEnd;
+        do {
+            _optimum[len--].Price = kIfinityPrice;
+        }
+        while (len >= 2);
+
+        for (i = 0; i < Base.kNumRepDistances; i++) {
+            int repLen = repLens[i];
+            if (repLen < 2) {
+                continue;
+            }
+            int price = repMatchPrice + getPureRepPrice(i, state, posState);
+            do {
+                int curAndLenPrice = price + _repMatchLenEncoder.getPrice(repLen - 2, posState);
+                Optimal optimum = _optimum[repLen];
+                if (curAndLenPrice < optimum.Price) {
+                    optimum.Price = curAndLenPrice;
+                    optimum.PosPrev = 0;
+                    optimum.BackPrev = i;
+                    optimum.Prev1IsChar = false;
+                }
+            }
+            while (--repLen >= 2);
+        }
+
+        int normalMatchPrice = matchPrice + com.swemel.sevenzip.compression.rangecoder.Encoder.getPrice0(_isRep[state]);
+
+        len = ((repLens[0] >= 2) ? repLens[0] + 1 : 2);
+        if (len <= lenMain) {
+            int offs = 0;
+            while (len > _matchDistances[offs]) {
+                offs += 2;
+            }
+            for (; ; len++) {
+                int distance = _matchDistances[offs + 1];
+                int curAndLenPrice = normalMatchPrice + getPosLenPrice(distance, len, posState);
+                Optimal optimum = _optimum[len];
+                if (curAndLenPrice < optimum.Price) {
+                    optimum.Price = curAndLenPrice;
+                    optimum.PosPrev = 0;
+                    optimum.BackPrev = distance + Base.kNumRepDistances;
+                    optimum.Prev1IsChar = false;
+                }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 654:659, 926:931.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.